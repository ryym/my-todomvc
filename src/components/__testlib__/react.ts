import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

// Though we used @testing-library/react first, it looks like it has a bug
// about the change event simulation (https://github.com/testing-library/react-testing-library/issues/359).
// And the official react-dom/test-utils is useful enough.

export const render = (element: ReactElement<any>) => {
  const container = document.createElement('div');
  act(() => {
    ReactDOM.render(element, container);
  });
  return container;
};

export const getByTestId = (container: HTMLElement, testId: string): HTMLElement => {
  const element = container.querySelector(`[data-testid="${testId}"]`) as HTMLElement;
  if (element == null) {
    throw new Error(`Test ID not found: ${testId}`);
  }
  return element;
};

export const renderSnapshot = (element: ReactElement<any>) => {
  return renderer.create(element).toJSON();
};
