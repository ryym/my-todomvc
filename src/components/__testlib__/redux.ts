import { Dispatch, Store, createStore } from 'redux';
import { State } from '../../store/state';

export function mockStore<S = State>(initialState: S, dispatch?: Dispatch): Store<S> {
  const noopReducer = (state: any) => state;
  let mockStore = createStore(noopReducer, initialState);
  if (dispatch) {
    mockStore = { ...mockStore, dispatch };
  }
  return mockStore;
}
