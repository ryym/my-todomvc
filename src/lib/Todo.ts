// XXX
let _globalId = 1;

export class Todo {
  static create(title: string, done: boolean = false) {
    return new Todo(_globalId++, title, done);
  }

  // XXX: Should be immutable?
  constructor(readonly id: number, public title: string, public done: boolean) {}
}
