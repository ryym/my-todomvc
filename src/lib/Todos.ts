export interface Todo {
  readonly id: number;
  readonly title: string;
  readonly completed: boolean;
}

export type Todos = Todo[];

export namespace Todos {
  export const init = (): Todos => {
    return [];
  };

  const _find = (todos: Todos, id: number): [Todo, number] => {
    for (let i = 0; i < todos.length; i++) {
      const todo = todos[i];
      if (todo.id === id) {
        return [todo, i];
      }
    }
    throw new Error(`Todo not found (id: ${id})`);
  };

  export const find = (todos: Todos, id: number): Todo => {
    const [todo] = _find(todos, id);
    return todo;
  };

  export const create = (id: number, title: string): Todo => {
    return { id, title, completed: false };
  };

  export const add = (todos: Todos, newTodo: Todo): Todos => {
    if (todos.filter(t => t.id === newTodo.id).length > 0) {
      throw new Error(`Todo (id: ${newTodo.id}) already exists`);
    }
    return [...todos, newTodo];
  };

  export const replace = (todos: Todos, updatedTodos: Todo[]): Todos => {
    todos = [...todos];
    updatedTodos.forEach(todo => {
      const [, idx] = _find(todos, todo.id);
      todos[idx] = todo;
    });
    return todos;
  };

  export const remove = (todos: Todos, ids: number[]): Todos => {
    return todos.filter(t => ids.indexOf(t.id) === -1);
  };
}
