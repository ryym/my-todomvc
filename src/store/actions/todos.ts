import { defineActions, effect } from 'redy';

let _globalTodoId = 1;

export const $todos = defineActions('todos', {
  AddNew: effect((title: string) => async dispatch => {
    if (title != null) {
      title = title.trim();
      if (title.length > 0) {
        dispatch($todos.Add(_globalTodoId++, title));
      }
    }
  }),

  Add: (id: number, title: string) => ({ id, title }),

  Remove: (id: number) => ({ id }),

  ChangeTitle: (id: number, title: string) => ({ id, title }),

  ToggleCompleted: (id: number, completed: boolean) => ({ id, completed }),

  ToggleCompletedAll: (completed: boolean) => ({ completed }),

  ClearCompleted: () => {},
});
