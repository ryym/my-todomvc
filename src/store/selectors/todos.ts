import { State } from '../state';
import { Todo } from '../../lib/Todos';
import { TodoState } from '../../lib/TodoState';
import { unreachable } from '../../lib/unreachable';

export const selectFilteredTodos = ({ todos, filters }: State): Todo[] => {
  switch (filters.todoState) {
    case TodoState.Value.All:
      return todos;

    case TodoState.Value.Active:
      return todos.filter(t => !t.completed);

    case TodoState.Value.Completed:
      return todos.filter(t => t.completed);

    default:
      return unreachable(filters.todoState);
  }
};
