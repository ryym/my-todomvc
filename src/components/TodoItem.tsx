import React, { KeyboardEvent, useState, useEffect, useRef } from 'react';
import { Dispatch } from 'redux';
import classNames from 'classnames';
import { $todos } from '../store/actions/todos';
import { Todo } from '../lib/Todos';

export type EditState = { type: 'EditStart'; id: number } | { type: 'EditEnd' };

export interface Props {
  readonly dispatch: Dispatch;
  readonly todo: Todo;
  readonly editable?: boolean;
  readonly onEditStateChange: (editState: EditState) => void;
}

export const TodoItem = ({
  dispatch,
  todo,
  editable = false,
  onEditStateChange,
}: Props) => {
  const titleInputRef = useRef<HTMLInputElement>(null);
  const [title, setTitle] = useState(todo.title);

  useEffect(() => {
    if (editable) {
      titleInputRef.current!.focus();
    }
  }, [editable]);

  const startEdit = () => {
    onEditStateChange({ type: 'EditStart', id: todo.id });
  };

  const finishEdit = () => {
    onEditStateChange({ type: 'EditEnd' });
  };

  const applyEdit = () => {
    if (title.length === 0) {
      dispatch($todos.Remove(todo.id));
    } else {
      dispatch($todos.ChangeTitle(todo.id, title));
    }
    finishEdit();
  };

  const titleEventHandlers = useTitleEventHandlers({
    editable,
    applyEdit,
    discardEdit: () => {
      setTitle(todo.title);
      finishEdit();
    },
  });

  return (
    <li className={classNames({ editing: editable, completed: todo.completed })}>
      <div className="view">
        <input
          className="toggle"
          type="checkbox"
          checked={todo.completed}
          onChange={event => {
            dispatch($todos.ToggleCompleted(todo.id, event.currentTarget.checked));
          }}
        />
        <label data-testid="title-label" onDoubleClick={startEdit}>
          {todo.title}
        </label>
        <button
          className="destroy"
          onClick={() => dispatch($todos.Remove(todo.id))}
        ></button>
      </div>
      <input
        data-testid="title-edit"
        ref={titleInputRef}
        type="text"
        value={title}
        className="edit"
        onChange={event => setTitle(event.currentTarget.value)}
        onKeyDown={titleEventHandlers.keyDown}
        onBlur={titleEventHandlers.blur}
      />
    </li>
  );
};

const useTitleEventHandlers = ({
  editable,
  applyEdit,
  discardEdit,
}: {
  editable: boolean;
  applyEdit: () => void;
  discardEdit: () => void;
}) => {
  const escKeyPressed = useRef<boolean>(false);

  const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    if (!editable) {
      return;
    }
    switch (event.key) {
      case 'Enter':
        applyEdit();
        break;
      case 'Esc':
      case 'Escape':
        // Call discardEdit here though we call it on blur too.
        // This ensures the item quits editable mode even if
        // some browser does not fire a blur event on Escape.
        discardEdit();
        escKeyPressed.current = true;
        break;
    }
  };

  const handleBlur = () => {
    // Delay the process to ensure that it runs after the keydown event handler.
    // This is because the blur and keydown events could runs in the different order
    // by the pressed key (at least in my environment).
    //   - When you exit by Tab - keydown -> blur
    //   - When you exit by Escape - blur -> keydown
    // And according to the spec, the changes made should be discarded
    // only when a user leave the editing mode by Escape.
    // (https://github.com/tastejs/todomvc/blob/master/app-spec.md#functionality)
    setTimeout(() => {
      if (escKeyPressed.current) {
        discardEdit();
        escKeyPressed.current = false;
      } else {
        applyEdit();
      }
    });
  };

  return { keyDown: handleKeyDown, blur: handleBlur };
};
