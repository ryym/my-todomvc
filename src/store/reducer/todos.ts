import { defineReducer, on } from 'redy';
import { $todos } from '../actions/todos';
import { Todos } from '../../lib/Todos';

export const reduceTodos = defineReducer(Todos.init(), [
  on($todos.Add, (todos, { id, title }) => {
    const newTodo = Todos.create(id, title);
    return Todos.add(todos, newTodo);
  }),

  on($todos.Remove, (todos, { id }) => {
    return Todos.remove(todos, [id]);
  }),

  on($todos.ChangeTitle, (todos, { id, title }) => {
    const todo = Todos.find(todos, id);
    const nextTodo = { ...todo, title };
    return Todos.replace(todos, [nextTodo]);
  }),

  on($todos.ToggleCompleted, (todos, { id, completed }) => {
    const todo = Todos.find(todos, id);
    return Todos.replace(todos, [{ ...todo, completed }]);
  }),

  on($todos.ToggleCompletedAll, (todos, { completed }) => {
    const toggledTodos = todos.map(todo => ({ ...todo, completed }));
    return Todos.replace(todos, toggledTodos);
  }),

  on($todos.ClearCompleted, todos => {
    const completedIds = todos.filter(todo => todo.completed).map(t => t.id);
    return Todos.remove(todos, completedIds);
  }),
]);
