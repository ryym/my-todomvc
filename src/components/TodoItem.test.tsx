import React from 'react';
import { SinonSpy, spy } from 'sinon';
import { act, Simulate } from 'react-dom/test-utils';
import { renderSnapshot, render, getByTestId } from './__testlib__/react';
import { Todo } from '../lib/Todos';
import { TodoItem } from './TodoItem';
import { $todos } from '../store/actions/todos';

it('matches snapshot', () => {
  const todo: Todo = { id: 1, title: 'todo1', completed: false };
  const tree = renderSnapshot(
    <TodoItem todo={todo} dispatch={spy()} onEditStateChange={spy()} />
  );
  expect(tree).toMatchSnapshot();
});

it('applies edited title on Enter', () => {
  const dispatch = spy();
  const onEditStateChange = spy();
  const todo: Todo = { id: 1, title: 'original title', completed: false };

  const container = render(
    <TodoItem
      editable
      todo={todo}
      dispatch={dispatch}
      onEditStateChange={onEditStateChange}
    />
  );

  const input = getByTestId(container, 'title-edit') as HTMLInputElement;
  act(() => {
    input.value = 'edited title';
    Simulate.change(input);
  });

  act(() => {
    Simulate.keyDown(input, { key: 'Enter' });
  });

  expect(dispatch.args).toEqual([[$todos.ChangeTitle(1, 'edited title')]]);
  expect(onEditStateChange.args).toEqual([[{ type: 'EditEnd' }]]);
});

it('discards edited title on Escape', async () => {
  const dispatch = spy();
  const onEditStateChange = spy();
  const todo: Todo = { id: 1, title: 'original title', completed: false };

  const container = render(
    <TodoItem
      editable
      todo={todo}
      dispatch={dispatch}
      onEditStateChange={onEditStateChange}
    />
  );

  const input = getByTestId(container, 'title-edit') as HTMLInputElement;
  act(() => {
    input.value = 'edited title';
    Simulate.change(input);
  });

  act(() => {
    Simulate.keyDown(input, { key: 'Escape' });
  });

  expect(dispatch.args).toEqual([]);
  expect(onEditStateChange.args).toEqual([[{ type: 'EditEnd' }]]);
});

it('applies edited title on blur', async () => {
  const dispatch = spy();
  const onEditStateChange = spy();
  const todo: Todo = { id: 1, title: 'original title', completed: false };

  const container = render(
    <TodoItem
      editable
      todo={todo}
      dispatch={dispatch}
      onEditStateChange={onEditStateChange}
    />
  );

  const input = getByTestId(container, 'title-edit') as HTMLInputElement;
  act(() => {
    input.value = 'edited title';
    Simulate.change(input);
  });

  act(() => {
    Simulate.blur(input);
  });
  await waitForFirstCall(onEditStateChange);

  expect(dispatch.args).toEqual([[$todos.ChangeTitle(1, 'edited title')]]);
});

const waitForFirstCall = (spy: SinonSpy, timeout: number = 1000): Promise<void> => {
  return new Promise((resolve, reject) => {
    const interval = 100;
    let elapsed = 0;
    const check = () => {
      if (spy.called) {
        resolve();
      } else if (elapsed < timeout) {
        elapsed += interval;
        setTimeout(check, interval);
      } else {
        reject(new Error(`Sinon function has not been called during ${elapsed}ms`));
      }
    };
    check();
  });
};
