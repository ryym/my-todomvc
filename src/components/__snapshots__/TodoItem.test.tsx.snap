// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<li
  className=""
>
  <div
    className="view"
  >
    <input
      checked={false}
      className="toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      data-testid="title-label"
      onDoubleClick={[Function]}
    >
      todo1
    </label>
    <button
      className="destroy"
      onClick={[Function]}
    />
  </div>
  <input
    className="edit"
    data-testid="title-edit"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="todo1"
  />
</li>
`;
